### PR TITLE
#303 Refactor ContainerCreator

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #303)
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

**Comment by chia7712** [comment]
FakeDockerClient has a bunch of code similar to ContainerCreator. This PR move all setter to ContainerCreator so other sub classes can reuse them.

**Comment by chia7712** [comment]
@saivirtue Could you file another PR to complete tests of TestContainerCreator if you have free cycle?

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala [hunk 1]
```diff
@@ -18,56 +18,103 @@ package com.island.ohara.agent.docker
...
+  protected def doExecute(hostname: String,
+                          imageName: String,
+                          name: String,
+                          command: String,
+                          removeContainerOnExit: Boolean,
+                          ports: Map[Int, Int],
+                          envs: Map[String, String],
+                          route: Map[String, String],
+                          volumeMapping: Map[String, String],
+                          networkDriver: NetworkDriver): Unit
```
Comment (saivirtue) [inline ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala]: I see...the point to reuse builder-pattern is put the "builder" part in interface and pass the builder params to the "override" sub-class, right?

**Comment by saivirtue** [comment]
> @saivirtue Could you file another PR to complete tests of TestContainerCreator if you have free cycle?

sure thing, I will add later.

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala (hunk 1 above)
Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala]: > the point to reuse builder-pattern is put the "builder" part in interface and pass the builder params to the "override" sub-class, right?

yep. However, this is just a variant of builder pattern. Not all use cases are suitable to this pattern :)

**Comment by chia7712** [comment]
Unrelated error. Will commit this PR

**Comment by saivirtue** [comment]
> Unrelated error. Will commit this PR

should I add commit to this branch?

**Comment by chia7712** [comment]
> should I add commit to this branch?

File another PR connected to #303 . Refactor is a big news worth having three pull requests :)

**Pull request merged**